pep-0690.rst

```diff
@@ -312,57 +325,54 @@ import object. When ``modb.foo`` is later referenced, it will also try to
-only into another dictionary) without being resolved. To preserve the
-performance of bulk-copy operations like ``dict.update()`` and ``dict.copy()``,
-they do not check for or resolve lazy import objects. However, if the source
-dict has a ``*_LAZY`` lookup kind set that indicates it might contain lazy
-objects, that lookup kind will be passed on to the updated/copied dictionary.
-This still ensures that the lazy import object can't escape into Python code
-without being resolved.
-
-Other "bulk" dictionary lookup methods (such as ``dict.items()``,
-``dict.values()``, etc) will resolve all lazy import objects in the dictionary.
-Since it is uncommon for any of these to be used on a module namespace
-dictionary, the priority here is simplicity of implementation and minimizing the
-overhead on normal non-lazy dictionaries (just one check to see if the
-dictionary has a ``*_LAZY`` lookup kind).
+There are two cases in which a lazy import object can "escape" one dictionary:
+
+* Into another dictionary: to preserve the performance of bulk-copy operations
+  like ``dict.update()`` and ``dict.copy()``, they do not check for or resolve
+  lazy import objects. However, if the source dict has the ``dk_lazy_imports``
+  flag set that indicates it might contain lazy objects, that flag will be
+  passed on to the updated/copied dictionary. This still ensures that the lazy
+  import object can't escape into Python code without being resolved.
+
+* Through the garbage collector: lazy imported objects are still Python objects
+  and live within the garbage collector; as such, they can be collected and seen
+  by means of using ``gc.collect()`` and ``gc.get_objects()``. Lazy objects are
+  not useful but they are also harmless and pose no danger if extracted from the
+  garbage collector in this way.
+
+Other "bulk" dictionary lookup methods involving values (such as ``dict.items()``,
+``dict.values()``, ``PyDict_Next()`` etc.) will attempt to resolve *all* lazy
+import objects in the dictionary prior to starting the iteration. Since only (some)
+module namespace dictionaries will ever have ``dk_lazy_imports`` set, the extra
+overhead of resolving all lazy import objects inside a dictionary is only paid
+by those dictionaries that need it. Minimizing the overhead on normal non-lazy
+dictionaries is the sole purpose of the ``dk_lazy_imports`` flag.
+
+``PyDict_Next`` will attempt to resolve all lazy import objects the first time
+position ``0`` is accessed, and those imports could fail with exceptions. Since
+``PyDict_Next`` cannot set an exception, ``PyDict_Next`` will return ``0``
+immediately in this case, and any exception will be printed to stderr as an
+unraisable exception.
+
+For this reason, this PEP introduces ``PyDict_NextWithError``, that works in the
+same way as ``PyDict_Next``, but which can set an error when returning ``0`` and
+this should be checked via ``PyErr_Occurred()`` after the call.
```

Comment:
I wonder if you could get away with not adding `PyDict_NextWithError` and just extend `PyDict_Next` with the semantics that `PyErr_Occurred()` should be checked after the call?  I think it's not possible for  `_PyDict_Next` (note the internal implementation function which `PyDict_Next` just wraps) to cause any other exceptions to occur so there wouldn't be any code currently checking for exceptions.

I wonder about this because of other APIs that could be iterating over dicts.  Now they have to think about whether the dict they iterate over could possibly have a lazy object in it, then they'd have to both change the function they call and check for the exception.  But how would they know whether the dict they are iterating over might or might not have lazy objects in them?

Reply:
In reality, we are not making use of `PyDict_NextWithError()` internally anywhere. This is added to make it easier to users to get real errors from resolving objects during `PyDict_Next()`, now that it can now throw errors related to imports or code execution during the resolution of lazy objects. If a user continues using `PyDict_Next()` any errors from the resolution of the lazy objects would only be printed and not raised.

`PyDict_Next()` already silently swallows what it probably should have been an exception, when the passed object is not a dictionary, and users could simply start using `PyDict_NextWithError()` always, no matter if they're expecting lazy objects or not.

Simply extending `PyDict_Next()` to need to always check for `PyErr_Occurred()` would produce problems for users of the API, because they don't expect it to throw any errors. I think we could maybe pave a way to go back to simply `PyDict_Next()` once users check for errors after the fact and don't assume the function can't throw any exceptions.

Reply:
I'll be interested in seeing what the SC and other community members think about this issue, so for now, it's find to publish the update with this proposal.  It seems to me that there will be a very long tail of `PyDict_Next` users and they will not be prepared to handle errors.  I think it would be helpful to provide some guidance for C API users as to when they will want to switch to `PyDict_NextWithError`.  Are you thinking that everyone who is iterating over a dict will eventually want to use the latter?  If you add this new API, then we'll have to live with it forever.

Reply:
Yes, I think the best would probably be for everyone to use `PyDict_NextWithError()` instead; much in the same way I believe `PyDict_GetItemWithError()` is preferred over `PyDict_GetItem()`. Generally, ignoring errors is not a good thing and right now passing a different (non-dictionary) object to `PyDict_Next()` ignores that and potentially hides problems. With Lazy Imports there will also be other kind of errors that could arise from iterating over a dict but I think the ignored type error alone is bad enough.